daemons/pacemakerd/pcmkd_subdaemons.c

```diff
@@ -127,44 +129,94 @@ pcmkd_cluster_connected(void)
 static gboolean
 check_active_before_startup_processes(gpointer user_data)
 {
-    gboolean keep_tracking = FALSE;
-
-    for (int i = 0; i < PCMK__NELEM(pcmk_children); i++) {
-        if (!pcmk_children[i].active_before_startup) {
-            /* we are already tracking it as a child process. */
-            continue;
-        } else {
-            int rc = child_liveness(&pcmk_children[i]);
-
-            switch (rc) {
-                case pcmk_rc_ok:
-                    break;
-                case pcmk_rc_ipc_unresponsive:
-                case pcmk_rc_ipc_pid_only: // This case: it was previously OK
-                    if (pcmk_children[i].respawn) {
-                        crm_err("%s[%lld] terminated%s", pcmk_children[i].name,
-                                (long long) PCMK__SPECIAL_PID_AS_0(pcmk_children[i].pid),
-                                (rc == pcmk_rc_ipc_pid_only)? " as IPC server" : "");
-                    } else {
-                        /* orderly shutdown */
-                        crm_notice("%s[%lld] terminated%s", pcmk_children[i].name,
-                                   (long long) PCMK__SPECIAL_PID_AS_0(pcmk_children[i].pid),
-                                   (rc == pcmk_rc_ipc_pid_only)? " as IPC server" : "");
-                    }
-                    pcmk_process_exit(&(pcmk_children[i]));
-                    continue;
-                default:
-                    crm_exit(CRM_EX_FATAL);
-                    break;  /* static analysis/noreturn */
+    static int next_child = 0;
+    int rc = child_liveness(&pcmk_children[next_child]);
+
+    crm_trace("%s[%lld] checked as %d",
+                           pcmk_children[next_child].name,
+                           (long long) PCMK__SPECIAL_PID_AS_0(
+                            pcmk_children[next_child].pid),
+                            rc);
+
+    switch (rc) {
+        case pcmk_rc_ok:
+            pcmk_children[next_child].check_count = 0;
+            next_child++;
+            subdaemon_check_progress = time(NULL);
+            break;
+        case pcmk_rc_ipc_pid_only: // This case: it was previously OK
+            pcmk_children[next_child].check_count++;
+            if (pcmk_children[next_child].check_count >= PCMK_PROCESS_CHECK_RETRIES) {
+                crm_err("%s[%lld] is unresponsive to ipc after %d tries but "
+                        "we found the pid so have it killed that we can restart",
+                        pcmk_children[next_child].name,
+                        (long long) PCMK__SPECIAL_PID_AS_0(
+                            pcmk_children[next_child].pid),
+                        pcmk_children[next_child].check_count);
+                stop_child(&pcmk_children[next_child], SIGKILL);
```

Comment:
In public clouds, nowadays it happens more often than before that sub-daemons are unresponsive to IPC and get respawned.

As we know, if it's controller that respawns, the node will lose all its transient attributes in the CIB status without being written again. Not only the resources that rely on the attributes will get impacted, but also missing of the internal attribute `#feature-set` will result into confusing `MIXED-VERSION` condition being shown from interfaces like crm_mon.

So far `PCMK_fail_fast=yes` probably is the only workaround to get the situation back into sanity but of course at a cost of node reboot.

While we've been trying to address it with the idea like:
https://github.com/ClusterLabs/pacemaker/pull/1699

, I'm not sure if it'd make sense at all to increase the tolerance here such as `PCMK_PROCESS_CHECK_RETRIES` or make it configurable... Otherwise should we say that 5 failures in a row are anyway bad enough to trigger a recovery?

Reply:
Sry I may be missing the reason for your comment here.
Previously IPC wasn't checked on a periodic basis for all subdaemons.
Numbers are kind of arbitrary. 1s is kind of a lower limit that makes sense for retries. Failing after 5  retries was the attempt to make it as reactive as before for cases where IPC was checked before already.

Reply:
Nothing is wrong with the changes in this PR. Just for bringing up the topic in the context here :-)

Reply:
Thanks for the info and opinion. I agree.